utils/utils.c

```diff
@@ -392,3 +392,95 @@ char * strjoin(char *left, char *right, char *delim)
 	strcpy(new + len - rlen - 1, right);
 	return new;
 }
+
+#define QUOTE '\''
+#define DQUOTE '"'
+
+static int setargs(char *args, char **argv)
+{
+	int count = 0;
+
+	/* ignore start spaces */
+	while (isspace(*args))
+		++args;
+
+	while (*args) {
+		/* consider quotes and update argv */
+		if (*args == QUOTE) {
+			if (argv)
+				argv[count] = ++args;
+			while (*args != QUOTE)
+				++args;
+			if (argv)
+				*args = ' ';
+		}
+		else if (*args == DQUOTE) {
+			if (argv)
+				argv[count] = ++args;
+			while (*args != DQUOTE)
+				++args;
+			if (argv)
+				*args = ' ';
+		}
+		else if (argv) {
+			argv[count] = args;
+		}
+		/* read characters until '\0' or space */
+		while (*args && !isspace(*args))
+			++args;
+		/* set '\0' rather than space */
+		if (argv && *args)
+			*args++ = '\0';
+		/* ignore margin spaces */
+		while (isspace(*args))
+			++args;
+		/* count up argument */
+		count++;
+	}
+
+	return count;
+}
+
+#undef QUOTE
+#undef DQUOTE
+
+char **parse_cmdline(char *cmd, int *argc)
```

Comment:
Maybe something like this:

```
/**
 * parse_cmdline - parse given string to be executed via execvp(3)
 * @cmd:  full command line
 * @argc: pointer to number of arguments
 *
 * This function parses @cmd and split it into an array of string
 * to be executed by exec(3) like argv[] in main().  The @argc
 * will be set to number of argument parsed if it's non-NULL.
 * The resulting array contains a copy of input string (@cmd) in
 * the first element which other elements point to.  It returns a
 * pointer to the second element so that it can be used directly
 * in other functions like exec(3).
 *
 * The returned array should be freed by free_parsed_cmdline().
 */
char **parse_cmdline(char *cmd, int *argc)